pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, timeAdded time.Time, reason string, consolidationType string) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates:        candidates,
+		Method:            reason,
+		ConsolidationType: consolidationType,
+		TimeAdded:         timeAdded,
+	}
+}
+
+func (q *Queue) Name() string {
+	return "disruption.queue"
+}
+
+func (q *Queue) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+func (q *Queue) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// Check if the queue is empty. client-go recommends not using this function to gate the subsequent
+	// get call, but since we're popping items off the queue synchronously, there should be no synchonization
+	// issues.
+	for q.Len() == 0 {
+		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+	// Get command from queue. This waits until queue is non-empty.
+	item, shutdown := q.RateLimitingInterface.Get()
+	if shutdown {
+		return reconcile.Result{}, fmt.Errorf("disruption queue has shut down")
+	}
+	cmd := item.(*Command)
+	defer q.RateLimitingInterface.Done(cmd)
+
+	if err := q.Process(ctx, cmd); err != nil {
+		// If recoverable, re-queue and try again.
+		if !IsUnrecoverableError(err) {
+			// log the error that is causing us to fail so we can bubble it up later if this times out.
+			cmd.LastError = err
+			q.RateLimitingInterface.AddRateLimited(cmd)
+			return reconcile.Result{RequeueAfter: controller.Immediately}, nil
+		}
+		// If the command failed, bail on the action.
+		// 1. Emit metrics for launch failures
+		// 2. Ensure cluster state no longer thinks these nodes are deleting
+		// 3. Remove it from the Queue's internal data structure
+		failedLaunches := lo.Filter(cmd.ReplacementKeys, func(key *NodeClaimKey, _ int) bool {
+			return !key.Initialized
+		})
+		disruptionReplacementNodeClaimFailedCounter.With(map[string]string{
+			methodLabel:            cmd.Method,
+			consolidationTypeLabel: cmd.ConsolidationType,
+		}).Add(float64(len(failedLaunches)))
+		q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+		multiErr := multierr.Combine(err, state.RequireNoScheduleTaint(ctx, q.kubeClient, false, cmd.Candidates...), cmd.LastError)
+
+		nodeNames := strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+			return s.Name()
+		}), ",")
+		q.Remove(cmd)
+		logging.FromContext(ctx).With("nodeNames", nodeNames).Errorf("failed to disrupt nodes, %w", multiErr)
+		return reconcile.Result{RequeueAfter: controller.Immediately}, nil
+	}
+	// If command is complete, remove command from queue.
+	q.Remove(cmd)
```

Comment:
Since we call q.Remove and the same reconcile result in both cases, should this all just be in a single return statement in the common code?

Reply:
The one difference is that we don't call q.Remove() when want to re-queue it, so I'm not sure that it makes sense to put these in a single return statement. What were you thinking here?

Reply:
We're already returning in the error case right up above on L167. Maybe I'm missing something, but I believe that's the requeue case so everything else should be the remove case (whether it's a unrecoverable error or a success)